examples/flutter_gallery/lib/demo/animation/home.dart

```diff
@@ -0,0 +1,533 @@
+
+import 'sections.dart';
+import 'widgets.dart';
+
+const Color _kAppBackgroundColor = const Color(0xFF353662);
+
+// This app's contents start out at _kHeadingMaxHeight and they function like
+// an appbar. Initially the appbar occupies most of the screen and its section
+// headings are laid out in a column. By the time its height has been
+// reduced to _kAppBarMidHeight, its layout is horizontal, only one section
+// heading is visible, and the section's list of details is visible below the
+// heading. The appbar's height can be reduced to no more than _kAppBarMinHeight.
+final double _kAppBarMinHeight = 90.0;
+final double _kAppBarMidHeight = 256.0;
+// The AppBar's max height depends on the screen, see AnimationDemoHomeState._buildBody()
+
+// Initially occupies the same space as the status bar and gets smaller as
+// the primary scrollable scrolls upwards.
+class _RenderStatusBarPaddingSliver extends RenderSliver {
+  _RenderStatusBarPaddingSliver({
+    double maxHeight,
+    double scrollFactor,
+  }) : _maxHeight = maxHeight, _scrollFactor = scrollFactor {
+    assert(maxHeight != null && maxHeight >= 0.0);
+    assert(scrollFactor != null && scrollFactor >= 1.0);
+  }
+
+  // The height of the status bar
+  double get maxHeight => _maxHeight;
+  double _maxHeight;
+  set maxHeight (double value) {
+    assert(maxHeight != null && maxHeight >= 0.0);
+    if (_maxHeight == value)
+      return;
+    _maxHeight = value;
+    markNeedsLayout();
+  }
+
+  // That rate at which this renderer's height shrinks when the scroll
+  // offset changes.
+  double get scrollFactor => _scrollFactor;
+  double _scrollFactor;
+  set scrollFactor (double value) {
+    assert(scrollFactor != null && scrollFactor >= 1.0);
+    if (_scrollFactor == value)
+      return;
+    _scrollFactor = value;
+    markNeedsLayout();
+  }
+
+  @override
+  void performLayout() {
+    final double height = (maxHeight - constraints.scrollOffset / scrollFactor).clamp(0.0, maxHeight);
+    geometry = new SliverGeometry(
+      paintExtent: math.min(height, constraints.remainingPaintExtent),
+      scrollExtent: maxHeight,
+      maxPaintExtent: maxHeight,
+    );
+  }
+}
+
+class _StatusBarPaddingSliver extends SingleChildRenderObjectWidget {
+  _StatusBarPaddingSliver({
+    Key key,
+    @required this.maxHeight,
+    this.scrollFactor: 5.0,
+  }) : super(key: key) {
+    assert(maxHeight != null && maxHeight >= 0.0);
+    assert(scrollFactor != null && scrollFactor >= 1.0);
+  }
+
+  final double maxHeight;
+  final double scrollFactor;
+
+  @override
+  _RenderStatusBarPaddingSliver createRenderObject(BuildContext context) {
+    return new _RenderStatusBarPaddingSliver(
+      maxHeight: maxHeight,
+      scrollFactor: scrollFactor,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderStatusBarPaddingSliver renderObject) {
+    renderObject.maxHeight = maxHeight;
+    renderObject.scrollFactor = scrollFactor;
```

Comment:
Nit: We usually use the `..` syntax because it's ever-so-slightly faster and these update functions are very hot.